muxers/mplex/src/io.rs

```diff
@@ -0,0 +1,771 @@
+        let max_buffer_len = config.max_buffer_len;
+        Multiplexed {
+            config,
+            status: Status::Ok,
+            io: Framed::new(io, Codec::new()).fuse(),
+            buffer: Vec::with_capacity(cmp::min(max_buffer_len, 512)),
+            open_substreams: Default::default(),
+            pending_flush: false,
+            pending_frames: Default::default(),
+            next_outbound_stream_id: LocalStreamId::dialer(0),
+            notifier_read: Arc::new(NotifierRead {
+                pending: Mutex::new(Default::default()),
+            }),
+            notifier_write: Arc::new(NotifierWrite {
+                pending: Mutex::new(Default::default()),
+            }),
+            notifier_open: Arc::new(NotifierOpen {
+                pending: Mutex::new(Default::default())
+            })
+        }
+    }
+
+    /// Flushes the underlying I/O stream.
+    pub fn poll_flush(&mut self, cx: &mut Context<'_>) -> Poll<io::Result<()>> {
+        match &self.status {
+            Status::Closed => return Poll::Ready(Ok(())),
+            Status::Err(e) => return Poll::Ready(Err(io::Error::new(e.kind(), e.to_string()))),
+            Status::Ok => {}
+        }
+
+        // Send any pending frames.
+        ready!(self.send_pending_frames(cx))?;
+
+        // Flush the underlying I/O stream.
+        let waker = NotifierWrite::register(&self.notifier_write, cx.waker());
+        match ready!(self.io.poll_flush_unpin(&mut Context::from_waker(&waker))) {
+            Err(e) => Poll::Ready(self.on_error(e)),
+            Ok(()) => {
+                self.pending_flush = false;
+                Poll::Ready(Ok(()))
+            }
+        }
+    }
+
+    /// Closes the underlying I/O stream.
+    ///
+    /// > **Note**: No `Close` or `Reset` frames are sent on open substreams
+    /// > before closing the underlying connection. However, the connection
+    /// > close implies a flush of any frames already sent.
+    pub fn poll_close(&mut self, cx: &mut Context<'_>) -> Poll<io::Result<()>> {
+        match &self.status {
+            Status::Closed => return Poll::Ready(Ok(())),
+            Status::Err(e) => return Poll::Ready(Err(io::Error::new(e.kind(), e.to_string()))),
+            Status::Ok => {}
+        }
+
+        // Note: We do not make the effort to send pending `Reset` frames
+        // here, we only close (and thus flush) the underlying I/O stream.
+
+        let waker = NotifierWrite::register(&self.notifier_write, cx.waker());
+        match self.io.poll_close_unpin(&mut Context::from_waker(&waker)) {
+            Poll::Pending => Poll::Pending,
+            Poll::Ready(Err(e)) => Poll::Ready(self.on_error(e)),
+            Poll::Ready(Ok(())) => {
+                self.pending_frames = VecDeque::new();
+                // We do not support read-after-close on the underlying
+                // I/O stream, hence clearing the buffer and substreams.
+                self.buffer = Vec::new();
```

Comment:
```suggestion
                self.buffer = Default::default();
```

Nit pick: To be consistent with `on_error`.